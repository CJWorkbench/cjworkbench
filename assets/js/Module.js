/**
 * A component that holds a single module that is then contained within the
 * Module Library.
 * The render function here will drive the "card" of each module within
 * the library.
 */

import React from 'react'
import PropTypes from 'prop-types'
import { CardBlock, Card } from 'reactstrap';

export default class Module extends React.Component {
  constructor(props) {
    super(props);
    this.initFields(props);
    this.state = {
      name: props.key,
      description: props.description,
      category: props.category,
      author: props.author,
    };
    this.itemClick = this.itemClick.bind(this);
    this.addModule = this.props.addModule.bind(this);
    this.workflow = this.props.workflow;
  }

  initFields(props) {
    this.key = props['data-name'];
    this.name = props['data-name'];
    this.description = props['data-description'];
    this.category = props['data-category'];
    this.author = props['data-author'];
    this.id = props['data-id'];
  }

  itemClick(evt) {
    var itemID = evt.target.getAttribute('data-id');
    this.props.addModule(this.props['data-id']);
    // Toggle temporarily disabled
    // this.workflow.toggleModuleLibrary();
  }

  render() {
    var moduleName = this.props['data-name'];
    var icon = 'icon-' + this.props['icon'] + ' ml-icon';

    return (
      // TODO: remove inline styles
      <div className='card' style={{'borderRadius': 0, 'border': 0}}>
        <div className='' onClick={this.itemClick} >
          <div className='second-level d-flex flex-row align-items-center'>
            <div className='ml-icon-container'>
              <div className={icon}></div>
            </div>
            <div>
              <div className='content-5 ml-module-name'>{moduleName}</div>
            </div>
          </div>
        </div>
      </div>
    )
  }
}

Module.propTypes = {
  addModule:  PropTypes.func,
  workflow: PropTypes.object,
};
