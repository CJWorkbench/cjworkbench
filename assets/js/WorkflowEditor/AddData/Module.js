import React from 'react'
import PropTypes from 'prop-types'

export default function Module ({ isLessonHighlight, idName, name, description, icon, onClick }) {
  const handleClick = React.useCallback(ev => { ev.preventDefault(); onClick(idName) })

  let className = 'module'
  if (isLessonHighlight) className += ' lesson-highlight'

  return (
    <a href='#' name={idName} onClick={handleClick} className={className}>
      <i className={`icon-${icon}`}/>
      <div className='text'>
        <h4>{name}</h4>
        <p>{description}</p>
      </div>
    </a>
  )
}
Module.propTypes = {
  idName: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired // func(idName) => undefined
}
