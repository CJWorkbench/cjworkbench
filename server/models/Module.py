# A Module defines a data processing action, with a set of paramters
# Module is the class, WfModule is the instance (applied in a Workflow)
from django.db import models

class Module(models.Model):
    # UI name, can change
    name = models.CharField('name', max_length=200)

    # internal name, cannot change if you want backwards compatibility with exported workflows
    id_name = models.CharField('id_name', max_length=200)

    # how do we run this module?
    dispatch = models.CharField('dispatch', max_length=200)

    def __str__(self):
        return self.name