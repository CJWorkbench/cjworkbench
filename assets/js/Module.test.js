/**
 * Testing Stories:
 * -Renders a Module card, with icon received from props
 * -Draggable feature exists
 * 
 * TODO:
 * 
 * -Not-read-only: calls AddModule from API on click
 * 
 * -Read-only state:
 *    -is not draggable
 *    -does not call AddModule
 * 
 */

import React from 'react'
import Module  from './Module'
import { mount } from 'enzyme'
import HTML5Backend from 'react-dnd-html5-backend'
import { DragDropContextProvider } from 'react-dnd'


describe('Module ', () => {
  
  var wrapper;  

  beforeEach(() => wrapper = mount(
    <DragDropContextProvider backend={HTML5Backend}>
      <Module
        key={"Sweet Module"}
        name={"Sweet Module"}
        icon={"add"}
        id={88}
        addModule={() => {}}
        dropModule={() => {}}
      />
    </DragDropContextProvider>
  ));
  afterEach(() => wrapper.unmount());

  it('Renders a card, with icon received from props', () => { 
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.icon-add')).toHaveLength(1);
  });

  it('Card is draggable', () => { 
    // search for property on the Module component that indicates drag-ability
    expect( Object.keys(wrapper.find('Module').props()).includes('connectDragSource') ).toBe(true);
  });
    
});