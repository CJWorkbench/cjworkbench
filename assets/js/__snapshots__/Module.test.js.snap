// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module NOT Read-only Renders snapshot 1`] = `
<div
  className="card ml-module-card "
  data-module-name="Sweet Module"
  onClick={[Function]}
>
  <div
    className="ML-module d-flex"
  >
    <div
      className="d-flex flex-row align-items-center"
    >
      <div
        className="ml-icon-container"
      >
        <div
          className="icon-add ml-icon"
        />
      </div>
      <div>
        <div
          className="content-5 ml-module-name"
        >
          Sweet Module
        </div>
      </div>
    </div>
    <div
      className="ml-handle"
    >
      <div
        className="icon-grip"
      />
    </div>
  </div>
</div>
`;
